06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+We show the plot of estimated accuracy versus the number of irrelevant predictors below.
+<center>
+```{r 06-performance-irrelevant-features, echo = FALSE, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Effect of inclusion of irrelevant predictors"}
+set.seed(1)
+# get accuracies after including k irrelevant features
+ks <- c(0, 5, 10, 20, 50, 100, 200, 500)
+accs <- list()
+for (i in 1:length(ks)) {
+  knn_spec <- nearest_neighbor(weight_func = "rectangular", neighbors = tune()) %>%
+     set_engine("kknn") %>%
+     set_mode("classification")
+
+  cancer_irrelevant_subset <- cancer_irrelevant %>% select(1:(3+ks[[i]]))
+
+  cancer_vfold <- vfold_cv(cancer_irrelevant_subset, v = 5, strata = Class)
+
+  cancer_recipe <- recipe(Class ~ ., data = cancer_irrelevant_subset) %>%
+      step_scale(all_predictors()) %>%
+      step_center(all_predictors())
+
+  
+  accs[[i]] <- workflow() %>%
+    add_recipe(cancer_recipe) %>%
+    add_model(knn_spec) %>%
+    tune_grid(resamples = cancer_vfold, grid = 10) %>%
+    collect_metrics() %>%
+    filter(.metric == "accuracy") %>%
+    summarize(mx = max(mean)) %>%
+    select(mx)
+  accs[[i]] <- accs[[i]]$mx
+}
+accs <- accs %>% unlist()
+
+# get accuracy if we always just guess the most frequent label
+base_acc <- cancer_irrelevant %>%
+                group_by(Class) %>%
+                summarize(n = n()) %>%
+                mutate(frac = n/sum(n)) %>%
+                summarize(mx = max(frac)) %>%
+                select(mx)
+base_acc <- base_acc$mx %>% unlist()
+
+# plot
+res <- tibble(ks = ks, accs = accs)
+res <- res %>% mutate(base_acc = base_acc)
+plt_irrelevant_accuracies <- res %>%
+  ggplot() +
+  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
+  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign (63%)")) +
+  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
+  scale_linetype_manual(name="Method", values = c("dashed", "solid"))
+
+plt_irrelevant_accuracies
+```
+</center>
+
+Wow! Once we add sufficiently many irrelevant predictor variables, our classifier
+does not do much better than just always guessing "benign", 
+ the most common class label in the data (63% of observations are benign).
+
+So then if we can't just throw in the kitchen sink, how do we choose which
+variables to use as predictors?  This is a very difficult question to answer in
+general, and there are a number of methods that have been developed that apply
+in particular cases of interest; see the additional resources at the end of
+this chapter to find out where you can learn more about variable selection.
+Here we will discuss *best subset selection,* which is applicable to any 
+classification method (K-NN or otherwise), and works when you have a small
+number of total variables from which you want to keep only the relevant ones (to
```

Comment:
What is a small number? Your example above had 502 predictors... Is that small? Would best subset work there? Maybe the explanation is just around the corner, but as someone who has never read this before I am wondering now.

Reply:
defined small as "at most 10" immediately